caustic/parameter.py

```diff
@@ -20,92 +21,69 @@ class Parameter:
     def __init__(
         self, value: Optional[Union[Tensor, float]] = None, shape: Optional[tuple[int, ...]] = ()
     ):
-        """
-        Initializes an instance of the Parameter class.
-
-        Args:
-            value (Optional[Union[Tensor, float]]: The value of the parameter. Defaults to None.
-            shape (Optional[tuple[int, ...]]): The shape of the parameter. Defaults to an empty tuple.
-
-        Raises:
-            ValueError: If both value and shape are None, or if shape is provided and doesn't match the shape of the value.
-        """
         # Must assign one of value or shape
-        self._value = value
         if value is None:
             if shape is None:
                 raise ValueError("If value is None, a shape must be provided")
+            if not isinstance(shape, tuple):
+                raise TypeError("The shape of a parameter must be a tuple")
             self._shape = shape
         else:
-            value = torch.as_tensor(value)
+            value = torch.as_tensor(value).float() # set float32 as default dtype here
             if shape != value.shape:
                 raise ValueError(
                     f"value's shape {value.shape} does not match provided shape {shape}"
                 )
-            self._value = value
             self._shape = value.shape
+        self._value = value
+        self._dtype = None if value is None else value.dtype
 
     @property
     def static(self) -> bool:
-        """
-        Checks if the parameter is static.
-
-        Returns:
-            bool: True if the parameter is static, False otherwise.
-        """
         return not self.dynamic
 
     @property
     def dynamic(self) -> bool:
-        """
-        Checks if the parameter is dynamic.
-
-        Returns:
-            bool: True if the parameter is dynamic, False otherwise.
-        """
         return self._value is None
 
     @property
     def value(self) -> Optional[Tensor]:
-        """
-        Returns the value of the parameter.
-
-        Returns:
-            Optional[Tensor]: The value of the parameter, or None if the parameter is dynamic.
-        """
         return self._value
+    
+    @value.setter
+    def value(self, value: Union[None, Tensor, float]):
+        if value is not None:
+            value = torch.as_tensor(value)
```

Comment:
are we going to enforce float32? if so this should be `torch.as_tensor(value, dtype = torch.float32)`

Reply:
I don't think we should enforce it in the value setter. After all, we might have a situation where float16 is set for the whole simulator and user pass a float16 value. Then enforcing float32 in the value setter will break the code in an unintuitive way.